File: Sources/AllUsage.swift
import Foundation
import CryptoKit

// Summaries only: persisted locally, never conversation text. Each session is read without descendants.
struct AllUsageSummary: Codable {
    var total = Tokens()
    var missing: [String: Int] = [:]
    var cost = APICost.unknown("正在汇总历史")
    var count = 0
    var loaded = 0
    var issues: [String] = []
    var updatedAt: String?
    func value(_ metric: Metric) -> String {
        if metric == .hitRate {
            return missing["input", default: 0] + missing["cached", default: 0] == 0 ? metric.value(total) : "—"
        }
        let text = metric.value(total)
        return text + (text != "—" && missing[metric.rawValue, default: 0] > 0 ? " + ?" : "")
    }
    var tokenLabel: String {
        guard let input = total.input, let output = total.output else { return "—" }
        let sum = input.addingReportingOverflow(output)
        guard !sum.overflow else { return "—" }
        return Metric.input.value(Tokens(input: sum.partialValue)) + (missing["input", default: 0] + missing["output", default: 0] > 0 ? " + ?" : "")
    }
    var status: String { if count == 0 && updatedAt == nil { return "正在读取本机历史…" }; return loaded < count ? "正在汇总 \(loaded) / \(count) 条记录" : "已汇总 \(count) 条记录 · 含归档与子代理" }
}

final class AllUsageReader {
    struct Cached: Codable { var signature: String; var tokens: Tokens?; var cost: APICost }
    private var cache: [String: Cached] = [:]
    private var entries: [ThreadEntry] = []
    private var pending: [ThreadEntry] = []
    private var current: (ThreadEntry, UsageReader)?
    private var refreshed = Date.distantPast
    private var issues: [String] = []
    private var updatedAt: String?
    private var dirty = false
    private var lastSaved = Date.distantPast
    private let cacheURL: URL?
    private let priceFingerprint: String
    private let catalog: () throws -> [ThreadEntry]
    init(cacheURL: URL? = Settings.directory.appendingPathComponent("all-usage-cache.json"), catalog: @escaping () throws -> [ThreadEntry] = { try UsageSources().allEntries() }) {
        self.cacheURL = cacheURL; self.catalog = catalog
        let encoder = JSONEncoder(); encoder.outputFormatting = [.sortedKeys]
        priceFingerprint = SHA256.hash(data: (try? encoder.encode(PriceBook.shared.document)) ?? Data()).description
        if let cacheURL, let data = try? Data(contentsOf: cacheURL), data.count <= 20 * 1024 * 1024 {
            cache = (try? JSONDecoder().decode([String: Cached].self, from: data)) ?? [:]
        }
    }
    private func key(_ e: ThreadEntry) -> String { e.source + ":" + e.id }
    private func signature(_ e: ThreadEntry) -> String {
        func stamp(_ path: String) -> String {
            guard let a = try? FileManager.default.attributesOfItem(atPath: path) else { return "missing" }
            return "\(a[.size] ?? "")|\(a[.modificationDate] ?? "")|\(a[.systemFileNumber] ?? "")"
        }
        return e.path + "|" + stamp(e.path) + (e.source == "opencode" ? stamp(e.path + "-wal") : "") + "|" + priceFingerprint
    }
    func poll(force: Bool = false) -> AllUsageSummary { autoreleasepool { pollBatch(force: force) } }
    private func pollBatch(force: Bool) -> AllUsageSummary {
        if current == nil && pending.isEmpty && (force || Date().timeIntervalSince(refreshed) >= 30) {
            refreshed = Date(); issues = []
            do {
                var seen: Set<String> = []
                entries = try catalog().filter { seen.insert(key($0)).inserted }
                let ids = Set(entries.map(key)); cache = cache.filter { ids.contains($0.key) }
                pending = entries.filter { cache[key($0)]?.signature != signature($0) }
                for e in pending { cache.removeValue(forKey: key(e)) }; dirty = true
            } catch { issues = ["部分数据源无法读取，汇总范围待核"] }
        }
        let deadline = Date().addingTimeInterval(0.15)
        repeat {
            if current == nil {
                guard !pending.isEmpty else { break }
                let e = pending.removeFirst()
                if e.source == "codex" { current = (e, UsageReader(path: e.path)) }
                else {
                    let before = signature(e)
                    let s = UsageSources().reader(for: e, includeChildren: false).poll()
                    dirty = true
                    cache[key(e)] = Cached(signature: before, tokens: s.total, cost: s.costTotal ?? .unknown("缺少费用记录"))
                    continue
                }
            }
            if let (e, reader) = current {
                let before = signature(e)
                let s = reader.poll()
                if reader.costs.hasMore { break }
                // If an actively written record changed, retry it next cycle rather than cache it as current.
                cache[key(e)] = Cached(signature: before, tokens: s.error == nil ? s.total : nil, cost: s.error == nil ? s.costTotal ?? .unknown("缺少费用记录") : .unknown("任务记录不可读"))
                current = nil; dirty = true
            }
        } while Date() < deadline
        if dirty && ((current == nil && pending.isEmpty) || Date().timeIntervalSince(lastSaved) >= 5) {
            dirty = false; lastSaved = Date()
            if current == nil && pending.isEmpty { updatedAt = ISO8601DateFormatter().string(from: Date()) }
            if let cacheURL, let data = try? JSONEncoder().encode(cache) {
                try? FileManager.default.createDirectory(at: cacheURL.deletingLastPathComponent(), withIntermediateDirectories: true)
                // Avoid rewriting an unchanged summary on every UI refresh.
                if (try? Data(contentsOf: cacheURL)) != data { try? data.write(to: cacheURL, options: .atomic) }
            }
        }
        return Self.summarize(entries: entries, cache: cache, issues: issues, updatedAt: updatedAt)
    }
    static func summarize(entries: [ThreadEntry], cache: [String: Cached], issues: [String] = [], updatedAt: String? = nil) -> AllUsageSummary {
        var s = AllUsageSummary(); s.issues = issues; s.updatedAt = updatedAt; s.cost = APICost()
        var seen: Set<String> = []
        let fields: [(String, WritableKeyPath<Tokens, Int64?>)] = [("input", \.input), ("output", \.output), ("cached", \.cached), ("written", \.written), ("reasoning", \.reasoning)]
        for e in entries where seen.insert(e.source + ":" + e.id).inserted {
            s.count += 1
            let row = cache[e.source + ":" + e.id]; if row != nil { s.loaded += 1 }
            s.cost = s.cost.adding(row?.cost ?? .unknown("历史尚未全部读取"))
            for (name, path) in fields {
                if let n = row?.tokens?[keyPath: path] {
                    let sum = (s.total[keyPath: path] ?? 0).addingReportingOverflow(n)
                    if sum.overflow { s.missing[name, default: 0] += 1 } else { s.total[keyPath: path] = sum.partialValue }
                } else { s.missing[name, default: 0] += 1 }
            }
        }
        if !issues.isEmpty { s.cost = s.cost.adding(.unknown(issues.joined(separator: "；"))); for (name, _) in fields { s.missing[name, default: 0] += 1 } }
        if s.count == 0 { s.cost = .unknown("未找到已接入的本地记录") }
        return s
    }
}

extension UsageSources {
    // No recent-menu limit or archive filter. Descendants are individual rows, so never recursively summed here.
    func allEntries(codexHome: URL? = nil, openCodePath: String? = nil) throws -> [ThreadEntry] {
        var result: [ThreadEntry] = []
        let home = codexHome ?? FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".codex")
        let databases = ((try? FileManager.default.contentsOfDirectory(at: home, includingPropertiesForKeys: nil)) ?? [])
            .filter { $0.lastPathComponent.range(of: "^state_[0-9]+\\.sqlite$", options: .regularExpression) != nil }
            .sorted { $0.lastPathComponent.localizedStandardCompare($1.lastPathComponent) == .orderedDescending }
        if let path = databases.first?.path {
            let rows = try UsageDatabase(path: path).rows("SELECT id,COALESCE(NULLIF(name,''),title) AS title,rollout_path FROM threads ORDER BY updated_at DESC")
            result += rows.map { ThreadEntry(id: $0["id"]!, title: $0["title"] ?? "", path: $0["rollout_path"] ?? "") }
        }
        let path = openCodePath ?? OpenCodeStore().path
        if FileManager.default.fileExists(atPath: path) {
            result += try UsageDatabase(path: path).rows("SELECT id,title FROM session").map { ThreadEntry(id: $0["id"]!, title: $0["title"] ?? "", path: path, source: "opencode") }
        }
        for (url, doc) in bridges {
            result += doc.sessions.map { ThreadEntry(id: $0.id, title: $0.title, path: url.path, source: "bridge:" + doc.id) }
        }
        return result
    }
}

File: windows/all_usage.py
"""All local sessions, including archived tasks and children, counted once per source/id."""
import json
import hashlib
import time
from pathlib import Path
import pricing
from usage_core import Catalog, FIELDS, MAX_INT, blank, now_iso, codex_log


def summary(entries=(), cache=None, issues=(), updated=None):
    cache=cache or {}; total=dict.fromkeys(FIELDS); missing=dict.fromkeys(FIELDS,0)
    result=dict(total=total,missing=missing,cost=pricing.cost(),count=0,loaded=0,issues=list(issues),updatedAt=updated)
    seen=set()
    for e in entries:
        key=e['source']+':'+e['id']
        if key in seen:continue
        seen.add(key);result['count']+=1;row=cache.get(key)
        if row is not None:result['loaded']+=1
        result['cost']=pricing.plus(result['cost'],row['cost'] if row else pricing.unknown('历史尚未全部读取'))
        for k in FIELDS:
            n=(row.get('tokens') or {}).get(k) if row else None
            if n is not None and 0 <= (total[k] or 0)+n <= MAX_INT:total[k]=(total[k] or 0)+n
            else:missing[k]+=1
    if issues:
        result['cost']=pricing.plus(result['cost'],pricing.unknown('部分数据源无法读取'))
        for k in FIELDS:missing[k]+=1
    if not result['count']:result['cost']=pricing.unknown('未找到已接入的本地记录')
    return result


def value(s,key):
    n=s['total'].get(key)
    return '—' if n is None else f'{n:,}'+(' + ?' if s['missing'].get(key,0) else '')


def token_label(s):
    i,o=s['total']['input'],s['total']['output']
    if i is None or o is None or i+o>MAX_INT:return '—'
    return f'{i+o:,}'+(' + ?' if s['missing']['input']+s['missing']['output'] else '')


def status(s):
    if not s['count'] and s['updatedAt'] is None:return '正在读取本机历史…'
    return f"正在汇总 {s['loaded']} / {s['count']} 条记录" if s['loaded']<s['count'] else f"已汇总 {s['count']} 条记录 · 含归档与子代理"


class AllUsageReader:
    def __init__(self,catalog=None,cache_path=None):
        self.catalog=catalog or Catalog();self.cache_path=cache_path;self.cache={}
        self.entries=[];self.pending=[];self.current=None;self.refreshed=float('-inf');self.issues=[];self.updated=None;self.dirty=False;self.last_saved=float('-inf')
        if cache_path:
            try:
                if cache_path.stat().st_size<=20*1024*1024:
                    data=json.loads(cache_path.read_text(encoding='utf-8'))
                    if isinstance(data,dict):self.cache=data
            except (OSError,ValueError):pass
        self.price_stamp=hashlib.sha256(json.dumps(pricing.BOOK.document,sort_keys=True).encode()).hexdigest()
    def key(self,e):return e['source']+':'+e['id']
    def signature(self,e):
        def stamp(p):
            try:
                s=Path(p).stat();return f'{s.st_size}|{s.st_mtime_ns}|{s.st_ino}'
            except OSError:return 'missing'
        return e['path']+'|'+stamp(e['path'])+('|' + stamp(e['path']+'-wal') if e['source']=='opencode' else '')+'|'+self.price_stamp
    def poll(self,force=False):
        if self.current is None and not self.pending and (force or time.monotonic()-self.refreshed>=30):
            self.refreshed=time.monotonic();self.issues=[]
            try:
                found=self.catalog.tasks(all_history=True,strict=True)
                self.entries=list({self.key(e):e for e in found}.values())
                self.cache={self.key(e):self.cache[self.key(e)] for e in self.entries if self.key(e) in self.cache}
                self.pending=[e for e in self.entries if self.cache.get(self.key(e),{}).get('signature')!=self.signature(e)]
                for e in self.pending:self.cache.pop(self.key(e),None)
                self.dirty=True
            except (OSError,ValueError,RuntimeError) as e:self.issues=['部分数据源无法读取，汇总范围待核']
        deadline=time.monotonic()+.15
        while time.monotonic()<deadline:
            if self.current is None:
                if not self.pending:break
                e=self.pending.pop(0)
                if e['source']=='codex':self.current=(e,pricing.CostLogReader(e['path']))
                else:
                    before=self.signature(e);s=self.catalog.snapshot(e,False)
                    self.cache[self.key(e)]=dict(signature=before,tokens=s['total'],cost=s.get('costTotal') or pricing.unknown('缺少费用记录'))
                    self.dirty=True;continue
            e,reader=self.current;before=self.signature(e);reader.poll()
            try:more=reader.offset<reader.path.stat().st_size
            except OSError:more=False
            if more:break
            s=codex_log(e['path'])
            self.cache[self.key(e)]=dict(signature=before,tokens=s['total'] if not s['error'] else None,cost=reader.value('total'))
            self.current=None;self.dirty=True
        if self.dirty and ((self.current is None and not self.pending) or time.monotonic()-self.last_saved>=5):
            self.dirty=False;self.last_saved=time.monotonic()
            if self.current is None and not self.pending:self.updated=now_iso()
            if self.cache_path:
                try:
                    self.cache_path.parent.mkdir(parents=True,exist_ok=True)
                    temp=self.cache_path.with_suffix('.tmp');temp.write_text(json.dumps(self.cache,ensure_ascii=False),encoding='utf-8');temp.replace(self.cache_path)
                except OSError:pass
        return summary(self.entries,self.cache,self.issues,self.updated)
